### ISSUE TITLE
Trouble Adding Rings to Polygons in Vector Layers: Works Inconsistently

### ISSUE THREAD

**Issue by Ipickedausername** [qgis/QGIS#59053]
### What is the bug or the crash?
Hello,

I am working with a SpatiaLite layer in CRS 2154 (the same as the project's CRS) with a Multi-polygon geometry type.
When I attempt to add a ring inside a polygon, it works inconsistently. 

Most of the time, if the process fails, creating another ring with fewer vertices and/or a smaller ring, will resolve the issue about 90% of the time.

I have tested this on other SpatiaLite layers with Polygon geometry, on Shapefile layers, and in different projects, but the issue persists.

### Steps to reproduce the issue

The issue is easy to reproduce; we simply need to add a ring with many vertices.
However, it seems very difficult to identify a pattern (number of vertices, shape of the ring, size of the ring).

Here are two pictures to show that the ring creation is possible within the same polygon, despite the error message.
![erreur](https://github.com/user-attachments/assets/d67d7b9f-c891-4139-9d34-2ca57dfb4326)
![paserreur](https://github.com/user-attachments/assets/0028fd0b-f80a-445d-b4c5-44b9cb9fda0f)

And, here is a test project where I can reproduce this issue. 
[Test_project.zip](https://github.com/user-attachments/files/17342620/Test_project.zip)

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

Version de QGIS | 3.38.3-Grenoble | Révision du code | 37f9e6efee
-- | -- | -- | --
Version de Qt | 5.15.13
Version de Python | 3.12.6
Version de GDAL/OGR | 3.9.2
Version de Proj | 9.4.0
Version de la base de données du registre EPSG | v11.004 (2024-02-24)
Version de GEOS | 3.12.2-CAPI-1.18.2
Version de SQLite | 3.45.1
Version de PDAL | 2.6.3
Version du client PostgreSQL | 16.2
Version de SpatiaLite | 5.1.0
Version de QWT | 6.2.0
Version de QScintilla2 | 2.14.1
Version de l'OS | Windows 10 Version 2009
  |   |   |  
Extensions Python actives
GroupStats | 2.2.7
mapswipetool_plugin | 1.3
networks | 2.7.19
OpenTopography-DEM-Downloader | 2.0
qfieldsync | v4.9.1
quick_map_services | 0.19.34
db_manager | 0.1.20
grassprovider | 2.12.99
processing | 2.12.99

</body></html>Version de QGIS
3.38.3-Grenoble
Révision du code
[37f9e6efee](https://github.com/qgis/QGIS/commit/37f9e6efee)
Version de Qt
5.15.13
Version de Python
3.12.6
Version de GDAL/OGR
3.9.2
Version de Proj
9.4.0
Version de la base de données du registre EPSG
v11.004 (2024-02-24)
Version de GEOS
3.12.2-CAPI-1.18.2
Version de SQLite
3.45.1
Version de PDAL
2.6.3
Version du client PostgreSQL
16.2
Version de SpatiaLite
5.1.0
Version de QWT
6.2.0
Version de QScintilla2
2.14.1
Version de l'OS
Windows 10 Version 2009

Extensions Python actives
GroupStats
2.2.7
mapswipetool_plugin
1.3
networks
2.7.19
OpenTopography-DEM-Downloader
2.0
qfieldsync
v4.9.1
quick_map_services
0.19.34
db_manager
0.1.20
grassprovider
2.12.99
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [x] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

**Comment by agiudiceandrea** [comment]
@Ipickedausername, thanks for reporting.

Are you sure the layer has the right CRS assigned? It looks like the features are located in the middle of the Atlantic Ocean and have a maximum dimension of few millimetres.

Does the issue also occur if you select the polygon feature before trying to add a ring in it?

The provided folder contains 4 layers and a project file and the provided project contains 3 layers. May you please describe which layer should be edited and how exactly in order to replicate the issue?

**Comment by Ipickedausername** [comment]
Hello,

Yes, I just created a project and specified the CRS before doing the tests. I didn't add a base map, so I didn't notice it was in the middle of the ocean. 

I adjusted the location of the polygons, somewhere in France, where Lambert93 apply. I was able to reproduce this issue. 

However, I wasn't able to reproduce the issue when the polygon is selected !

So, it seems that, if the polygon isn't selected, QGIS cannot always detect where the ring is located?
